core/src/cgcloud/core/box.py

```diff
@@ -365,15 +370,86 @@ def prepare( self, ec2_keypair_globs, instance_type=None, image_ref=None,
+        if options['price'] != None:
+            assert spot_support is True
+            # this means we are using a spot bid. We need a couple additional kwargs
+            self._optimize_bid(instance_type, options, kwargs)
+
+            kwargs['price']=options['price']
+
+            kwargs['launch_group']=self.ctx.namespace.replace("/","")+"_launch_group"
+            try: # There is no min/max count parameter for request_spot_instances, just count
+                kwargs['count']=kwargs['max_count']
+                del kwargs['max_count']
+                del kwargs['min_count']
+            except KeyError:
+                pass
         self.instance_creation_args = kwargs
 
+    def _choose_zone(self, bid, spot_data):
+        # returns the zone to put your spot request based on:
+        #    1) zones with prices currently under the bid
+        #    2) zones with the most stable price
+
+        ZoneTuple = namedtuple("ZoneTuple", ["name","price_deviation"])
+        def _sort_zones(bid, spot_data, zones):
+            # returns two lists of tuples of form: [ (zone.name, std_deviation)...]
+            # one for zones over the bid price and one for zones under bid price. Each are sorted by
+            # increasing standard deviation values
+            markets_under_bid, markets_over_bid = [], []
+            for zone in zones:
+                zlist = filter(lambda spot_history: spot_history.availability_zone == zone.name, spot_data)
+                std_deviation = numpy.std([spot_history.price for spot_history in zlist])
+                # http://stackoverflow.com/a/12135169
+                (markets_over_bid, markets_under_bid)[spot_data[0].price < bid].append(ZoneTuple(zone.name,std_deviation))
+            # sort each list in place by increasing standard deviation
+            map(lambda list: list.sort(key=lambda zone_tuple:zone_tuple.price_deviation), [markets_over_bid, markets_under_bid])
+            return markets_under_bid, markets_over_bid
+
+        markets_under_bid, markets_over_bid = _sort_zones(bid, spot_data, self.ctx.ec2.get_all_zones())
+        most_stable_market = markets_under_bid[0].name if markets_under_bid else markets_over_bid[0].name
+        return most_stable_market
+
+    def _optimize_bid(self, instance_type, options, kwargs):
+        # insures that the bid is an appropriate price and makes an effort to place it in a sensible zone
+        bid = float(options['price'])
+        spot_data = self._get_spot_history(instance_type)
+        self._check_bid(bid, spot_data)
+        most_stable_zone = self._choose_zone(bid,spot_data)
+        kwargs["placement"]= most_stable_zone
+        log.info("Requesting spot instances in zone %s", most_stable_zone)
+
+    def _check_bid(self, bid, spotData):
+        # Don't allow crazy bids.
+        average = numpy.mean([datum.price for datum in spotData])
+        try:
+            assert (bid < (average * 2))
```

Comment:
Raise a UserError instead of this. UserError exceptions are printed without a stack trace.